lib/Runtime/Library/JavascriptRegExpConstructor.cpp

```diff
@@ -53,17 +54,44 @@ namespace Js
         this->lastInput = lastInput;
         this->lastMatch = lastMatch;
         this->reset = true;
+        this->invalidatedLastMatch = false;
+    }
+
+    void JavascriptRegExpConstructor::InvalidateLastMatch(UnifiedRegex::RegexPattern* lastPattern, JavascriptString* lastInput)
+    {
+        AssertMsg(lastPattern != nullptr, "lastPattern should not be null");
+        AssertMsg(lastInput != nullptr, "lastInput should not be null");
+        AssertMsg(JavascriptOperators::GetTypeId(lastInput) != TypeIds_Null, "lastInput should not be JavaScript null");
+
+        this->lastPattern = lastPattern;
+        this->lastInput = lastInput;
+        this->lastMatch.Reset();
+        this->reset = true;
```

Comment:
Does "reset" in the two lines above have the same meaning?

Reply:
No. We want to reset the last match to undefined and let the regex constructor know it has to reset it's cache via `JavascriptRegExpConstructor::EnsureValues`.

Reply:
What is the difference between reset and invalidateLastMatch?

Reply:
The existing flag `reset` is used to indicate we should reset the cache backed-by `JavascriptRegExpConstructor::EnsureValues`. I added `invalidateLastMatch` which is used to indicate that there is no value stored in the last match field. This means we need to first calculate the last match before we use it. Ordinarily we assert that the last match field in `JavascriptRegExpConstructor::EnsureValues` is not undefined.

Reply:
The reset flag materializes all of the RegExp['$blah'] properties as actual JavascriptStrings (when needed) rather than input + offset which is faster to compute, and the logic ensures this only happens once. I'm not sure how the term "reset" applies to this logic but this code is apparently very old.